test/functional/utils/performance.py

```diff
@@ -0,0 +1,216 @@
+
+    def __str__(self):
+        return f"p{self.value:g}".replace(".", "_")
+
+
+class IOMetric(ValidatableParameter):
+    read_IOPS = Schema(Use(int))
+    write_IOPS = Schema(Use(int))
+    read_BW = Schema(Use(int))
+    write_BW = Schema(Use(int))
+    read_CLAT_AVG = Schema(Use(int))
+    write_CLAT_AVG = Schema(Use(int))
+    read_CLAT_PERCENTILES = Schema({Use(PercentileMetric): Use(int)})
+    write_CLAT_PERCENTILES = Schema({Use(PercentileMetric): Use(int)})
+
+BuildTypes = ["master", "pr", "other"]
+
+class ConfigParameter(ValidatableParameter):
+    CAS_VERSION = Schema(Use(str))
+    DUT = Schema(Use(str))
+    TEST_NAME = Schema(str)
+    BUILD_TYPE = Schema(Or(*BuildTypes))
+    CACHE_CONFIG = Schema(
+        {"cache_mode": Use(str), "cache_line_size": Use(str), "cleaning_policy": Use(str)}
+    )
+    CACHE_TYPE = Schema(Use(str))
+    CORE_TYPE = Schema(Use(str))
+    TIMESTAMP = Schema(And(datetime, Use(str)))
+
+
+class WorkloadParameter(ValidatableParameter):
+    NUM_JOBS = Schema(Use(int))
+    QUEUE_DEPTH = Schema(Use(int))
+
+
+class MetricContainer:
+    def __init__(self, metric_type):
+        self.metrics = {}
+        self.metric_type = metric_type
+
+    def insert_metric(self, metric, kind):
+        if not isinstance(kind, self.metric_type):
+            raise Exception(f"Invalid metric type. Expected: {self.metric_type}, got: {type(kind)}")
+
+        if kind.value:
+            metric = kind.value.validate(metric)
+
+        self.metrics[kind] = metric
+
+    @property
+    def is_empty(self):
+        return len(self.metrics) == 0
+
+    def to_serializable_dict(self):
+        # No easy way for json.dump to deal with custom classes (especially custom Enums)
+        def stringify_dict(d):
+            new_dict = {}
+            for k, v in d.items():
+                k = str(k)
+
+                if isinstance(v, dict):
+                    v = stringify_dict(v)
+                elif isinstance(v, int):
+                    pass
+                elif isinstance(v, float):
+                    pass
+                else:
+                    v = str(v)
+
+                new_dict[k] = v
+
+            return new_dict
+
+        return stringify_dict(self.metrics)
+
+
+class PerfContainer:
+    def __init__(self):
+        self.conf_params = MetricContainer(ConfigParameter)
+
+        self.workload_params = MetricContainer(WorkloadParameter)
+
+        self.cache_metrics = MetricContainer(IOMetric)
+        self.core_metrics = MetricContainer(IOMetric)
+        self.exp_obj_metrics = MetricContainer(IOMetric)
+
+    def insert_config_param(self, param, kind: ConfigParameter):
+        self.conf_params.insert_metric(param, kind)
+
+    def insert_config_from_cache(self, cache):
+        cache_config = {
+            "cache_mode": cache.get_cache_mode(),
+            "cache_line_size": cache.get_cache_line_size(),
+            "cleaning_policy": cache.get_cleaning_policy(),
+        }
```

Comment:
I'm wondering if information about promotion policy and sequential cutoff would be wort to keep

Reply:
For now we don't plan on testing performance with those settings. So in spirit of YAGNI I didn't include it.